confgenerator/apache.go

```diff
@@ -0,0 +1,167 @@
+		LoggingProcessorParseShared: LoggingProcessorParseShared{
+			TimeKey:    "time",
+			TimeFormat: "%d/%b/%Y:%H:%M:%S %z",
+			Types: map[string]string{
+				"http_request_status": "integer",
+				// N.B. "http_request_responseSize" is a string containing an integer.
+				// https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#HttpRequest.FIELDS.response_size
+			},
+		},
+	}.Components(tag, uid)
+	// apache logs "-" when a field does not have a value. Remove the field entirely when this happens.
+	for _, field := range []string{
+		"host",
+		"user",
+		"http_request_referer",
+	} {
+		c = append(c, fluentbit.Component{
+			Kind: "FILTER",
+			Config: map[string]string{
+				"Name":      "modify",
+				"Match":     tag,
+				"Condition": fmt.Sprintf("Key_Value_Equals %s -", field),
+				"Remove":    field,
+			},
+		})
+	}
+	// Generate the httpRequest structure.
+	c = append(c, fluentbit.Component{
+		Kind: "FILTER",
+		Config: map[string]string{
+			"Name":          "nest",
+			"Match":         tag,
+			"Operation":     "nest",
+			"Wildcard":      "http_request_*",
+			"Nest_under":    "logging.googleapis.com/http_request",
+			"Remove_prefix": "http_request_",
+		},
+	})
+	return c
+}
+
+type LoggingProcessorApacheError struct {
+	ConfigComponent `yaml:",inline"`
+}
+
+func (LoggingProcessorApacheError) Type() string {
+	return "apache_error"
+}
+
+func (p LoggingProcessorApacheError) Components(tag string, uid string) []fluentbit.Component {
+	c := LoggingProcessorParseRegex{
+		// Documentation: https://httpd.apache.org/docs/current/logs.html#errorlog
+		// Sample line 2.4: [Fri Sep 09 10:42:29.902022 2011] [core:error] [pid 35708:tid 4328636416] (13)Permission denied [client 72.15.99.187] File does not exist: /usr/local/apache2/htdocs/favicon.ico
+		// Sample line 2.2: [Fri Sep 09 10:42:29.902022 2011] [error] [pid 35708:tid 4328636416] [client 72.15.99.187] File does not exist: /usr/local/apache2/htdocs/favicon.ico
+		// TODO - Support time parsing for version 2.0 where smallest resolution is seconds
+		// Sample line 2.0: [Wed Oct 11 14:32:52 2000] [error] [client 127.0.0.1] client denied by server configuration: /export/home/live/ap/htdocs/test
+		Regex: `^\[(?<time>[^\]]+)\] \[(?:(?<module>\w+):)?(?<level>[\w\d]+)\](?: \[pid (?<pid>\d+)(?::tid (?<tid>[0-9]+))?\])?(?: (?<errorCode>[^\[]*))?(?: \[client (?<client>[^\]]*)\])? (?<message>.*)$`,
+		LoggingProcessorParseShared: LoggingProcessorParseShared{
+			TimeKey:    "time",
+			TimeFormat: "%a %b %d %H:%M:%S.%L %Y",
+			Types: map[string]string{
+				"pid": "integer",
+				"tid": "integer",
+			},
+		},
+	}.Components(tag, uid)
+	for _, l := range []struct{ level, severity string }{
+		{"emerg", "EMERGENCY"},
+		{"alert", "ALERT"},
+		{"crit", "CRITICAL"},
+		{"error", "ERROR"},
+		{"warn", "WARNING"},
+		{"notice", "NOTICE"},
+		{"info", "INFO"},
+		{"debug", "DEBUG"},
+		{"trace1", "DEBUG"},
```

Comment:
Trace8 is the maximum. Simply for simplicity of the code I opted to not make a separate key_value_matches block, but we could add it in instead of the current pattern.

Reply:
Got it. Can we add a comment to mention the rationale? Optionally also add a link to the Apache documentation reference that Trace8 is the maximum (if there is a link)?

Reply:
Added a comment & doc link, let me know if you have any further thoughts on it.

Reply:
Perfect. Thanks!